applications/FluidDynamicsApplication/custom_elements/d_vms_dem_coupled.cpp

```diff
@@ -402,42 +497,228 @@ void DVMSDEMCoupled<TElementData>::CalculateStabilizationParameters(
+    const array_1d<double,Dim>& r_predicted_subscale = mPredictedSubscaleVelocity[rData.IntegrationPointIndex];
+
+    for (unsigned int d = 0; d < Dim; d++) {
+        convective_velocity[d] += r_predicted_subscale[d];
+    }
+
+    return convective_velocity;
+}
+
+template< class TElementData >
+void DVMSDEMCoupled<TElementData>::UpdateSubscaleVelocityPrediction(
+    const TElementData& rData)
+{
+    const double density = this->GetAtCoordinate(rData.Density,rData.N);
+    const double viscosity = this->GetAtCoordinate(rData.EffectiveViscosity,rData.N);
+    array_1d<double,3> resolved_convection_velocity = this->GetAtCoordinate(rData.Velocity,rData.N) - this->GetAtCoordinate(rData.MeshVelocity,rData.N);
+
+    const double dt = rData.DeltaTime;
+    const double h = rData.ElementSize;
+
+    // Elemental large-scale velocity gradient
+    BoundedMatrix<double,Dim,Dim> resolved_velocity_gradient = ZeroMatrix(Dim,Dim);
+
+    const auto& r_resolved_velocities = rData.Velocity;
+    for (unsigned int i = 0; i < NumNodes; i++) {
+        for (unsigned int m = 0; m < Dim; m++) {
+            for (unsigned int n = 0; n < Dim; n++) {
+                resolved_velocity_gradient(m,n) += rData.DN_DX(i,n) * r_resolved_velocities(i,m);
+            }
+        }
+    }
+
+    const array_1d<double,Dim>& old_subscale_velocity = mOldSubscaleVelocity[rData.IntegrationPointIndex];
 
-    double det_inv_tau = MathUtils<double>::Det(inv_tau);
-    MathUtils<double>::InvertMatrix(inv_tau, non_diag_tau_one, det_inv_tau, -1.0);
+    // Part of the residual that does not depend on the subscale
+    array_1d<double,3> static_residual = ZeroVector(3);
 
-    MathUtils<double>::GaussSeidelEigenSystem<BoundedMatrix<double,Dim,Dim>, BoundedMatrix<double,Dim,Dim>>(non_diag_tau_one, eigen_vectors_matrix, eigen_values_matrix);
+    // Note I'm only using large scale convection here, small-scale convection is re-evaluated at each iteration.
+    if (rData.UseOSS != 1.0)
+        this->AlgebraicMomentumResidual(rData,resolved_convection_velocity,static_residual);
+    else
+        this->OrthogonalMomentumResidual(rData,resolved_convection_velocity,static_residual);
+
+    // Add the time discretization term to obtain the part of the residual that does not change during iteration
+    for (unsigned int d = 0; d < Dim; d++)
+        static_residual[d] += density/dt * old_subscale_velocity[d];
+
+    constexpr double c1 = DVMS<TElementData>::mTauC1;
+    constexpr double c2 = DVMS<TElementData>::mTauC2;
+    constexpr static double subscale_prediction_velocity_tolerance = DVMS<TElementData>::mSubscalePredictionVelocityTolerance;
+    constexpr static double subscale_prediction_maximum_iterations = DVMS<TElementData>::mSubscalePredictionMaxIterations;
+    constexpr static double subscale_prediction_residual_tolerance = DVMS<TElementData>::mSubscalePredictionResidualTolerance;
```

Comment:
You can define these in the type definition section of the header so you don't need to redefined them again in the other places where they are used.